modules/spring-config/src/main/java/org/mule/runtime/config/internal/model/ApplicationModel.java

```diff
@@ -331,35 +323,67 @@ public ApplicationModel(ArtifactConfig artifactConfig, ArtifactDeclaration artif
+        return topLevelComponentsStream()
+            .flatMap(cm -> cm.recursiveStream());
+      }
+
+      @Override
+      public Spliterator<ComponentAst> recursiveSpliterator() {
+        return recursiveStream().spliterator();
+      }
+
+      @Override
+      public Stream<ComponentAst> topLevelComponentsStream() {
+        return muleComponentModels.stream();
+      }
+
+      @Override
+      public Spliterator<ComponentAst> topLevelComponentsSpliterator() {
+        return topLevelComponentsStream().spliterator();
+      }
+    };
+
     createEffectiveModel();
     indexComponentModels();
     validateModel(componentBuildingDefinitionRegistry);
     extensionModelHelper = new ExtensionModelHelper(extensionModels);
     // TODO MULE-13894 do this only on runtimeMode=true once unified extensionModel names to use camelCase (see smart connectors
     // and crafted declared extension models)
     resolveComponentTypes();
+    resolveMetadataTypes(extensionModelHelper, recursiveStream());
     topLevelComponentsStream()
         .forEach(componentModel -> ((ComponentModel) componentModel).resolveTypedComponentIdentifier(extensionModelHelper,
                                                                                                      runtimeMode));
     recursiveStreamWithHierarchy(this).forEach(new ComponentLocationVisitor());
   }
 
+  /**
+   * We force the current instance of {@link ApplicationModel} to be highly cohesive with {@link MacroExpansionModulesModel} as
+   * it's responsibility of this object to properly initialize and expand every global element/operation into the concrete set of
+   * message processors
+   *
+   * @param extensionModels Set of {@link ExtensionModel extensionModels} that will be used to check if the element has to be
+   *        expanded.
+   * @param postProcess a closure to be executed after the macroexpansion of an extension.
+   */
   public void macroExpandXmlSdkComponents(Set<ExtensionModel> extensionModels) {
-    expandModules(extensionModels, () -> {
-      // TODO MULE-13894 do this only on runtimeMode=true once unified extensionModel names to use camelCase (see smart connectors
-      // connectors and crafted declared extension models)
-
-      // TODO MULE-17419 (AST) Remove these 2 actions
-      resolveComponentTypes();
-      topLevelComponentsStream()
-          .forEach(componentModel -> ((ComponentModel) componentModel).resolveTypedComponentIdentifier(extensionModelHelper,
-                                                                                                       runtimeMode));
+    final ArtifactAst previousAst = ast;
+    ast = new MacroExpansionModulesModel(previousAst, extensionModels).expand();
 
+    if (ast != previousAst) {
       // Have to index again the component models with macro expanded ones
-      indexComponentModels();
-    });
+      // TODO MULE-17197 (AST) uncomment this
+      // indexComponentModels();
```

Comment:
Meantime, would this break anything? I guess it was just a matter of performance to have them indexed, right?

Reply:
from what i've seen, the maps populated by this method is used in the dsl processing. Now, there is another question: why does the dsl processing of an app care about its macroexpanded components? Maybe this method did something else once and now it makes no sense to call it here?